gh-125916: Adapt functools.reduce() to Argument Clinic

skirpichev opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-125916 -->
* Issue: gh-125916
<!-- /gh-issue-number -->

skirpichev committed:
gh-125916: convert reduce() to Argument Clinic

skirpichev committed:
+1

JelleZijlstra commented on Modules/_functoolsmodule.c:
```diff
@@ -932,15 +932,31 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
+
+    function as func: object
+    iterable as seq: object
+    initial as result: object(c_default="NULL") = _functools._initial_missing
```
Do we really need to add this `_initial_missing` object?

skirpichev replied:
I see no other options, just as in the pure-Python version.  What you could suggest instead?  ``None`` and ``...`` were rejected.

If this is no-go - perhaps reduce() shouldn't be converted to AC, until it get support for multiple signatures.

JelleZijlstra replied:
An alternative is to write the c_default as NULL and manually write an `@text_signature`. This is what is done by `_collections.deque.index`.

skirpichev replied:
Ah, I see, that's an option.  The drawback is that the inspect.signature() can't parse this and pure-Python version will have a different signature.

I'll postpone this change till @erlend-aasland review.

erlend-aasland replied:
Yes, my gut feel when I first browsed this PR earlier today, was also to use `@text_signature`. I think we should consider it. I'm not sure how relevant differences between the C extension module and the Python version of `functools` are.

serhiy-storchaka commented on Modules/_functoolsmodule.c:
```diff
@@ -932,15 +932,31 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
+
+    function as func: object
+    iterable as seq: object
+    initial as result: object(c_default="NULL") = _functools._initial_missing
```
Why not simply use NULL?
```suggestion
    initial as result: object = NULL
```

erlend-aasland commented on Modules/_functoolsmodule.c:
```diff
@@ -932,15 +932,31 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
+a default when the iterable is empty.
+
+For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5])
+calculates ((((1 + 2) + 3) + 4) + 5).
```
Let's keep the docstring as it was. If it needs amendment, we should do so in a separate PR, because _that_ PR would then be backported to all active branches (though, of course backported manually bco. the Argument Clinic changes).

skirpichev replied:
Unfortunately, AC requires a summary line.  The reduce() docstring doesn't fit in that format as-is.

erlend-aasland replied:
Right, so let's amend the docstring first (in a separate PR).

JelleZijlstra replied:
The docstring change here seems small enough that it's fine to do it also in this PR.

skirpichev replied:
FYI: https://github.com/python/cpython/pull/126045

skirpichev replied:
pr updated to include docstrings changes.

skirpichev committed:
address review: make signature unrepresentable again

skirpichev committed:
+1

JelleZijlstra commented on Modules/_functoolsmodule.c:
```diff
@@ -932,15 +932,32 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
-// Not converted to argument clinic, because of `args` in-place modification.
-// AC will affect performance.
+/*[clinic input]
+@text_signature "($module, function, iterable[, initial], /)"
```
With Serhyi's suggestion we don't actually need `@text_signature` any more. If you remove it, you get:

```
>>> functools.reduce.__text_signature__
'($module, function, iterable, initial=<unrepresentable>, /)'
```

There is an open issue #87233 about supporting such signatures in inspect. I suggest we remove this line for now so that the signature becomes representable once that issue is fixed.
```suggestion
```

skirpichev replied:
I'm -0 on this - lets reviewers decide if this suggestion should be applied.

To me, using ``@text_signature`` basically means: FIXME, we actually can't represent signature for this function.  I think that in most (all?) cases fixing this will require support for multiple signatures.  So, the clinic input must be manually modified.  And ``@text_signature`` marks are easily greppable.

erlend-aasland replied:
I'm fine with both Serhiy's suggestion and `@text_signature`. Since both Jelle and Alex is in favour of Serhiy's suggestion, let's go for that.

skirpichev replied:
Alex?

erlend-aasland replied:
> Alex?

Yes, he reacted with a thumbs-up to Jelle's https://github.com/python/cpython/pull/125999#discussion_r1817966240

skirpichev committed:
Improve functools.reduce() docstring

PEP 257 says that "Multi-line docstrings consist of a summary line just
like a one-line docstring, followed by a blank line, followed by a more
elaborate description."  This is also a requirement from the Argument
Clinic.

skirpichev committed:
Apply suggestions from code review

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

skirpichev committed:
Merge branch 'main' into reduce-docstring

skirpichev commented on Modules/_functoolsmodule.c:
```diff
@@ -932,15 +932,32 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
+
+Apply a function of two arguments cumulatively to an iterable, from left to right.
+
+This efficiently reduces the iterable to a single value.  If initial is present,
```
```suggestion
This effectively reduces the iterable to a single value.  If initial is present,
```

skirpichev committed:
Update Modules/_functoolsmodule.c

skirpichev committed:
Update Lib/functools.py

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

erlend-aasland committed:
Update Modules/_functoolsmodule.c

skirpichev committed:
Merge branch 'reduce-docstring' into reduce-to-ac

skirpichev committed:
Merge branch 'master' into reduce-to-ac

skirpichev committed:
+ clinic

erlend-aasland committed:
Merge branch 'main' into reduce-to-ac

skirpichev committed:
Update Modules/_functoolsmodule.c

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

skirpichev committed:
+1

serhiy-storchaka commented on Modules/_functoolsmodule.c:
```diff
@@ -933,28 +933,27 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
 
 For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5])
-calculates ((((1 + 2) + 3) + 4) + 5).
+calculates ((((1 + 2) + 3) + 4) + 5).")
```
```suggestion
calculates ((((1 + 2) + 3) + 4) + 5).
```

skirpichev committed:
Update Modules/_functoolsmodule.c

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

skirpichev committed:
+1
